packages/view3d-optimizer/index.js

```diff
@@ -0,0 +1,322 @@
+    const pngImages = images
+      .filter(img => img.mimeType ? img.mimeType === "image/png" : img.uri.endsWith("png"))
+      .map(img => path.resolve(tempDir.path, img.uri));
+
+    const commonConfig = `--resize "{\"width\": ${minSize}, \"height\": ${minSize}}" --suffix "_${minSize}" --output-dir "${tempDir.path}"`;
+
+    const tasks = [];
+    if (jpgImages.length > 0) tasks.push(`npx squoosh-cli --mozjpeg auto ${commonConfig} ${jpgImages.join(" ")}`);
+    if (pngImages.length > 0) tasks.push(`npx squoosh-cli --oxipng auto ${commonConfig} ${pngImages.join(" ")}`);
+
+    if (tasks.length > 0) {
+      await run(tasks, `Applying texture LOD (${minSize})`);
+
+      for (const texture of textures) {
+        const origImage = images[texture.source];
+        const origSize = await probe(fs.createReadStream(path.resolve(tempDir.path, origImage.uri)));
+        const newTextureIndex = gltf.textures.length;
+
+        // Copy current texture
+        gltf.textures.push({
+          ...texture
+        });
+
+        texture.source = gltf.images.length;
+
+        if (!texture.extras) texture.extras = {};
+
+        texture.extras["view3d-lod"] = {
+          levels: [{
+            size: Math.max(origSize.width, origSize.height),
+            index: newTextureIndex
+          }]
+        };
+
+        const newImage = {};
+
+        if (origImage.mimeType) newImage.mimeType = origImage.mimeType;
+
+        const extName = path.extname(origImage.uri);
+        const fileName = path.basename(origImage.uri, extName);
+
+        newImage.uri = `${fileName}_${minSize}${extName}`;
+
+        gltf.images.push(newImage);
+      };
+
+      // Save modified gltf
+      fs.writeJSONSync(tmpFile, gltf);
+    }
+  }
+
+  if (options.textureLod1) {
+    const thresholdSize = parseFloat(options.textureLod1);
+    if (isNaN(thresholdSize)) {
+      console.error(chalk.red("Failed to parse the texture-lod-1 texture size"));
+      process.exit(1);
+    }
+
+    const gltf = fs.readJSONSync(tmpFile);
+    const textures = [...gltf.textures];
+    const images = textures
+      .map(texture => gltf.images[texture.source])
+      .filter(img => !!img.uri);
+    const jpgImages = [];
+    const pngImages = [];
+
+    const commonConfig = `--resize "{\"width\": ${thresholdSize}, \"height\": ${thresholdSize}}" --suffix "_${thresholdSize}" --output-dir "${tempDir.path}"`;
+
+    for (const texture of textures) {
+      const origImage = images[texture.source];
+      const origSize = await probe(fs.createReadStream(path.resolve(tempDir.path, origImage.uri)));
+
+      if (Math.max(origSize.width, origSize.height) <= thresholdSize) break;
+
+      const newTextureIndex = gltf.textures.length;
+
+      gltf.textures.push({
+        sampler: texture.sampler,
+        source: newTextureIndex
+      });
+
+      if (!texture.extras) texture.extras = {};
+      if (!texture.extras["view3d-lod"]) texture.extras["view3d-lod"] = [];
```

Comment:
I saw the code that uses the extra in view3d, it looks like an object containing an array of `levels`.

Reply:
Oh well you're right, it should be an empty object 👍